### PR TITLE
Refactor `mean()` reducer to use `FExpr`

### PR DESCRIPTION
**Pull request by samukweku**
- refactor `mean()` reducer to use `FExpr`;
- fix `median()` to use `float64` for void columns.

WIP for #2562

### CODE REVIEW

**Comment by samukweku** [comment]
@oleksiyskononenko I do not understand the rules for `void` when computing the mean. If you have some time, I'd love your feedback on it

File: src/core/expr/fexpr_mean.cc [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        gby = Groupby::single_group(wf.nrows());
+      } 
+
+      for (size_t i = 0; i < wf.ncols(); ++i) {
+      bool is_grouped = ctx.has_group_column(
+                        wf.get_frame_id(i),
+                        wf.get_column_id(i)
+                      );
+      Column coli = evaluate1(wf.retrieve_column(i), gby, is_grouped);        
+      outputs.add_column(std::move(coli), wf.retrieve_name(i), Grouping::GtoONE);
```
Comment (oleksiyskononenko) [inline src/core/expr/fexpr_mean.cc:68]:
```suggestion
        bool is_grouped = ctx.has_group_column(
                            wf.get_frame_id(i),
                            wf.get_column_id(i)
                          );
        Column coli = evaluate1(wf.retrieve_column(i), gby, is_grouped);        
        outputs.add_column(std::move(coli), wf.retrieve_name(i), Grouping::GtoONE);         
```

File: src/core/column/mean.h [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        col_(std::move(col)),
+        gby_(gby),
+        is_grouped_(is_grouped)
+    {
+      xassert(col_.can_be_read_as<T>());
+    }
+
+
+    bool get_element(size_t i, T* out) const override {
+
```
Comment (oleksiyskononenko) [inline src/core/column/mean.h:49]:
```suggestion
```

File: src/core/expr/fexpr_mean.cc [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+      switch (stype) {
+        case SType::VOID: {
+          return Column(new ConstNa_ColumnImpl(gby.size(), stype));
+        }
+          
+        case SType::BOOL:
+        case SType::INT8:
+        case SType::INT16:
+        case SType::INT32:        
+        case SType::INT64:
```
Comment (oleksiyskononenko) [inline src/core/expr/fexpr_mean.cc:87]: I don't think you can just fall through here. You can't `get_element()` from the int columns to `double`. See how a similar `switch` was implemented before.

File: src/core/expr/fexpr_mean.cc [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        case SType::FLOAT64:
+          return make<double>(std::move(col), SType::FLOAT64, gby, is_grouped);
+        case SType::TIME64: {
+          Column coli = make<double>(std::move(col), SType::FLOAT64, gby, is_grouped);
+          coli.cast_inplace(SType::TIME64);
+          return coli;
+        }
```
Comment (oleksiyskononenko) [inline src/core/expr/fexpr_mean.cc:94]: We can probably do a similar trick to `DATE32`. Also, if we support datetimes, we need to adjust documentation — it now says we only support numerics for `mean()`.

**Comment by oleksiyskononenko** [comment]
@samukweku So what is your question on voids?

File: src/core/expr/fexpr_mean.cc [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+      return outputs;
+    }
+
+
+    Column evaluate1(Column &&col, const Groupby& gby, bool is_grouped) const {
+      SType stype = col.stype();
+
+      switch (stype) {
+        case SType::VOID: {
+          return Column(new ConstNa_ColumnImpl(gby.size(), stype));
```
Comment (oleksiyskononenko) [inline src/core/expr/fexpr_mean.cc:80]:
```suggestion
          return Column(new ConstNa_ColumnImpl(gby.size(), SType::FLOAT64));
```

Comment (oleksiyskononenko) [reply src/core/expr/fexpr_mean.cc:80]: According to documentation
>The column stypes are float32 for float32 columns, and float64 for all the other numeric types.

That's the reason you need to return `SType::FLOAT64` column, and not `SType::VOID`.

File: src/core/expr/fexpr_mean.cc (hunk 3 above)
Comment (samukweku) [reply src/core/expr/fexpr_mean.cc:87]:
```
using U = typename std::conditional<std::is_same<T, float>::value,
                                      float, double>::type;
```

this is the switch condition used. not familiar with the construct - my interpretation is that if the type is float32, then the output is float32, else the output should be float64. that's why i implemented it as such. open to learning

Comment (oleksiyskononenko) [reply src/core/expr/fexpr_mean.cc:87]: Yes, you're correct with respect to the return type. I didn't notice you're casting in-place in the `make()` function. It's all ok, I guess.

File: src/core/expr/fexpr_mean.cc [hunk 6]
```diff
@@ -64,20 +66,23 @@ class FExpr_Mean : public FExpr_Func {
...
+    Column evaluate1(Column &&col, const Groupby& gby, bool has_by, bool is_grouped) const {
       SType stype = col.stype();
 
       switch (stype) {
         case SType::VOID: {
-          return Column(new ConstNa_ColumnImpl(gby.size(), SType::FLOAT64));
+          std::cout<<has_by<<std::endl;
+          Column coli = has_by? Column(new ConstNa_ColumnImpl(gby.size(), SType::FLOAT64))
+                              : Column(new ConstNa_ColumnImpl(gby.size()));
+          return coli;
```
Comment (oleksiyskononenko) [inline src/core/expr/fexpr_mean.cc:85]: This seems to be some debug changes?

File: src/core/expr/fexpr_mean.cc [hunk 7]
```diff
@@ -64,20 +66,23 @@ class FExpr_Mean : public FExpr_Func {
                             wf.get_frame_id(i),
                             wf.get_column_id(i)
                           );
-        Column coli = evaluate1(wf.retrieve_column(i), gby, is_grouped);        
+        Column coli = evaluate1(wf.retrieve_column(i), gby, has_by, is_grouped);
```
Comment (oleksiyskononenko) [inline src/core/expr/fexpr_mean.cc:68]: So why do we need to have `has_by`? The code was just fine before this commit, it was only the stype that was wrong in several tests.

Comment (samukweku) [reply src/core/expr/fexpr_mean.cc:68]: I guess I misunderstood then. So all void columns should return float64, whether or not a groupby is present?

Comment (oleksiyskononenko) [reply src/core/expr/fexpr_mean.cc:68]: Yeah, there is no reason to distinguish.

**Comment by oleksiyskononenko** [comment]
Is there anything else you're planning to do on this PR? Not sure why the build is failing.

**Comment by samukweku** [comment]
@oleksiyskononenko Just ur review and merge is what's left

**Comment by oleksiyskononenko** [comment]
The build is read, so we first need to make it green before merging. Can you re-trigger the build to see if it was a temp glitch on AppVeyor?

**Comment by samukweku** [comment]
Sure, will do

**Comment by oleksiyskononenko** [comment]
@samukweku once I merge https://github.com/h2oai/datatable/pull/3412, please remove your changes to `test_median*` test and merge main onto your PR. `test_median*` tests have typos and testing `mean()` instead of `median()`.

**Comment by oleksiyskononenko** [comment]
Nevermind, I've already made the required fixes. See if you have anything else to be done for this PR, otherwise, I guess it is all good to be merged.

**Comment by samukweku** [comment]
Thanks @oleksiyskononenko all good from me. Ok to merge

**Review by oleksiyskononenko — APPROVED** [review]

**Pull request merged**